lib/kernel.js

```diff
@@ -66,65 +118,122 @@ export default class Kernel {
   }
 
   addWatchCallback(watchCallback: Function) {
-    this.watchCallbacks.push(watchCallback);
-  }
-
-  _callWatchCallbacks() {
-    this.watchCallbacks.forEach(watchCallback => watchCallback());
+    this.transport.addWatchCallback(watchCallback);
   }
 
   interrupt() {
-    throw new Error("Kernel: interrupt method not implemented");
+    this.middlewareStack.interrupt();
   }
 
   shutdown() {
-    throw new Error("Kernel: shutdown method not implemented");
+    this.middlewareStack.shutdown();
   }
 
   restart(onRestarted: ?Function) {
-    throw new Error("Kernel: restart method not implemented");
+    this.middlewareStack.restart(onRestarted);
   }
 
   execute(code: string, onResults: Function) {
-    throw new Error("Kernel: execute method not implemented");
+    this.middlewareStack.execute(
+      code,
+      true,
+      this._wrapExecuteResultsCallback(onResults)
+    );
   }
 
   executeWatch(code: string, onResults: Function) {
-    throw new Error("Kernel: executeWatch method not implemented");
+    this.middlewareStack.execute(
+      code,
+      false,
+      this._wrapExecuteResultsCallback(onResults)
+    );
+  }
+
+  _wrapExecuteResultsCallback(onResults: Function) {
```

Comment:
The name of this function how it does what it does, but it doesn't say what it does).
Please, add the JSDoc documentation for method and describe its purpose.